### ISSUE TITLE
Certificate Expiry Notification is broken after upgrade to 1.20

### ISSUE THREAD

**Issue by StaceZ** [openxpki/openxpki#631]
My notify_expiry.yaml does not found any certificates after upgrade from 1.19 to 1.20.

If server started with debug: "**openxpkictl start --debug OpenXPKI::Server::Workflow::Activity::Tools::SearchCertificates:64**" I see:

---
2018-03-28 15:53:16.681858 DEBUG:32 PID:4112 OpenXPKI::Server::Workflow::Activity::Tools::SearchCertificates::execute (line 159): Full query $VAR1 = {
          'CERT_ATTRIBUTES' => [],
          'ENTITY_ONLY' => 1,
          'NOTAFTER' => {
                          'OPERATOR' => 'BETWEEN',
                          'VALUE' => [
                                       1522227196,
                                       1523436796
                                     ]
                        },
          'NOTBEFORE' => {
                           'OPERATOR' => 'LESS_THAN',
                           'VALUE' => 1522227196
                         },
          'STATUS' => 'ISSUED'
        };

2018-03-28 15:53:17.261367 DEBUG:64 PID:4112 OpenXPKI::Server::Workflow::Activity::Tools::SearchCertificates::execute (line 162): Search returned $VAR1 = [];

---

But "**openxpkiadm certificate list --realm ca-NPK -all -v -v -v**" show certificate expired soon:

---

 Identifier: G4EGUuv1CujY6zysB9K0uW7C0B0
    Subject:      UID=planshet+CN=fda,DC=npk,DC=ru
    Issuer DN:      emailAddress=hostmaster@npk.ru,CN=NPK CONTACT Registration Authority,OU=IT,O=NPK CONTACT,L=Novosibirsk,ST=Novosibirskaya Oblast,C=RU
    Chain:      G4EGUuv1CujY6zysB9K0uW7C0B0 -> YpTRZvamM-NMJNlIt_FVeGBAEH0 -> JPUHlj0bqlqReF2O0lcRgB6TnWE(complete)
    subject key identifier:      F1:B7:75:53:F1:F5:F6:B6:54:66:8B:DB:F7:36:F0:12:B5:93:5B:31
    authority key identifier:      63:AA:EB:CF:6D:09:78:7D:29:CD:9E:1E:66:0D:A2:1E:CD:DF:9A:DE
    cert key:      412369357278824652630783
    issuer identifier:      YpTRZvamM-NMJNlIt_FVeGBAEH0
    status:      ISSUED
    notafter:      1522310840
    notbefore:      1506672440
    req key:      22527

---

**Comment by StaceZ** [comment]
I think the error somewhere in the API ;(
'NOTAFTER' in JSON query converted to "certificate.notbefore" in SQL ;(((

Details: 
1) notify_expiry.yaml:
```initialize:
        class: OpenXPKI::Server::Workflow::Activity::Tools::SearchCertificates
        param:
          entity_only: 1
          target_key: tmp_queue
          cutoff_notafter: +000030
```
2) Full query $VAR1 = {
```          'CERT_ATTRIBUTES' => [],
          'ENTITY_ONLY' => 1,
          'NOTAFTER' => {
                          'OPERATOR' => 'BETWEEN',
                          'VALUE' => [
                                       1521549059,
                                       1530966659
                                     ]
                        },
          'NOTBEFORE' => {
                           'OPERATOR' => 'LESS_THAN',
                           'VALUE' => 1522413059
                         },
          'STATUS' => 'ISSUED'
        };
```
3) Generated MySQL query:
SELECT certificate.* FROM certificate WHERE ( ( ( certificate.notbefore BETWEEN '1522419581' AND '1525011583' ) AND certificate.pki_realm = 'ca-NPK' AND certificate.req_key IS NOT NULL AND certificate.status = 'ISSUED' ) ) ORDER BY certificate.cert_key DESC

**Comment by maxhq** [comment]
Hi StaceZ, I will try to reproduce your problem. But the timestamps in your full query output do not match the ones in the DB query. Could you please double check you got the right query (or is it fron two different runs)?

**Comment by StaceZ** [comment]
Yes data from different runs. Sorry.
Data from single new run:
```
Full query $VAR1 = {
          'CERT_ATTRIBUTES' => [],
          'ENTITY_ONLY' => 1,
          'NOTAFTER' => {
                          'OPERATOR' => 'BETWEEN',
                          'VALUE' => [
                                       1522646656,
                                       1525238656
                                     ]
                        },
          'NOTBEFORE' => {
                           'OPERATOR' => 'LESS_THAN',
                           'VALUE' => 1522646656
                         },
          'STATUS' => 'ISSUED'
        };
```


SELECT certificate.* FROM certificate WHERE ( ( ( certificate.notbefore BETWEEN '1522646655' AND '1525238657' ) AND certificate.pki_realm = 'ca-NPK' AND certificate.req_key IS NOT NULL AND certificate.status = 'ISSUED' ) ) ORDER BY certificate.cert_key DESC

**Comment by StaceZ** [comment]
Hi! Have you tried it? Have any results?
If not, then, apparently, I had to go back to the previous version.
Where can I download 1.19 for Ubintu?
In the repository of this version is no longer, and I did not guess to save the distribution; (

**Comment by maxhq** [comment]
Hi and sorry for the delay. We have backported some fixes from the 2.x version branch.
This should also fix your problems.
A new release 1.20.2 is in preparation.

**Comment by StaceZ** [comment]
Thank you! I'll wait. Can you name the approximate date of occurrence 1.20.2 on Ubuntu 14.04?

**Comment by oliwel** [comment]
I put 1.20.2 packages on the server but did not test it - please give it a try.

**Comment by StaceZ** [comment]
Unfortunately, I do not see any new packages for Ubuntu. Only packages for Debian are updated ;(

**Comment by oliwel** [comment]
Sorry - I have created and uploaded ubuntu packages now, please try them

**Comment by StaceZ** [comment]
Certificate expiration notifications now work!
Thank you!

In addition, now loading a username into the UTF8 from LDAP does not cause the web interface to crash when viewing workflows.
Thanks again!

A couple of small comments:
1) I have HTML files already in Unicode, so I had to edit the file /usr/lib/perl5/OpenXPKI/Server/Notification / SMTP.pm by removing "Encode :: encode_utf8 (...)" from it to avoid excessive transcoding.

I think it would be better to make a configuration parameter indicating the need for such a recoding.

2) the logic was inverted:
```
condition:
    is_renewed:
        class: OpenXPKI :: Server :: Workflow :: Condition :: WFArray
        param:
            array_name: renewed_identifier
            condition: is_empty
```

Previously, if this array was empty, then is_renewed had a true value, which would spoil its name, but it corresponds to the specified condition.
Therefore, I renamed this variable to is_not_renewed and changed the conditions at the beginning of the file.
Now, if no new certificates are found, then is_renewed has the value of a false, although, as far as I understand, this contradicts the given conditions.
